File: server/routes/recommend.js
const express = require('express');
const router = express.Router();
const { execFile } = require('child_process');
const path = require('path');

// POST /api/recommend — Get personalized recommendations
router.post('/', (req, res) => {
  const { name, dob, goals, budget, colorPreference } = req.body;

  if (!dob || !goals || !budget) {
    return res.status(400).json({
      success: false,
      error: 'Missing required fields: dob, goals, budget'
    });
  }

  const input = JSON.stringify({ name, dob, goals, budget, colorPreference });
  const pythonScript = path.join(__dirname, '..', 'python', 'recommend.py');

  execFile('python', [pythonScript, input], { timeout: 10000 }, (error, stdout, stderr) => {
    if (error) {
      // Fallback: try python3 command
      execFile('python3', [pythonScript, input], { timeout: 10000 }, (error2, stdout2, stderr2) => {
        if (error2) {
          console.error('Python error:', stderr || stderr2);
          // Ultimate fallback: use JS-based recommendation
          try {
            const result = jsRecommend(req.body);
            return res.json({ success: true, data: result });
          } catch (jsErr) {
            return res.status(500).json({
              success: false,
              error: 'Recommendation engine failed'
            });
          }
        }
        try {
          const result = JSON.parse(stdout2);
          res.json({ success: true, data: result });
        } catch (parseErr) {
          res.status(500).json({ success: false, error: 'Failed to parse recommendations' });
        }
      });
      return;
    }

    try {
      const result = JSON.parse(stdout);
      res.json({ success: true, data: result });
    } catch (parseErr) {
      res.status(500).json({ success: false, error: 'Failed to parse recommendations' });
    }
  });
});

// JavaScript fallback recommendation engine
function jsRecommend(input) {
  const fs = require('fs');
  const dataPath = path.join(__dirname, '..', 'data', 'gemstones.json');
  const gemstones = JSON.parse(fs.readFileSync(dataPath, 'utf-8'));

  const { dob, goals, budget, colorPreference } = input;

  // Calculate zodiac sign
  const date = new Date(dob);
  const month = date.getMonth() + 1;
  const day = date.getDate();
  const zodiac = getZodiacSign(month, day);

  // Score each gemstone
  const scored = gemstones.map(gem => {
    let score = 0;

    // Zodiac match (40%)
    if (gem.zodiacSigns.map(z => z.toLowerCase()).includes(zodiac.toLowerCase())) {
      score += 40;
    }

    // Goals match (25%)
    if (goals && goals.length > 0) {
      const matchedGoals = goals.filter(g => gem.goals.includes(g.toLowerCase()));
      score += (matchedGoals.length / goals.length) * 25;
    }

    // Budget match (20%)
    if (gem.priceRange === budget) {
      score += 20;
    } else if (
      (budget === 'mid' && gem.priceRange === 'affordable') ||
      (budget === 'premium' && gem.priceRange === 'mid')
    ) {
      score += 10;
    }

    // Color preference (15%)
    if (colorPreference && gem.color.toLowerCase() === colorPreference.toLowerCase()) {
      score += 15;
    }

    return { ...gem, matchScore: Math.round(score) };
  });

  scored.sort((a, b) => b.matchScore - a.matchScore);

  return {
    zodiacSign: zodiac,
    recommendations: scored.slice(0, 3)
  };
}

function getZodiacSign(month, day) {
  const signs = [
    { sign: 'Capricorn', start: [1, 1], end: [1, 19] },
    { sign: 'Aquarius', start: [1, 20], end: [2, 18] },
    { sign: 'Pisces', start: [2, 19], end: [3, 20] },
    { sign: 'Aries', start: [3, 21], end: [4, 19] },
    { sign: 'Taurus', start: [4, 20], end: [5, 20] },
    { sign: 'Gemini', start: [5, 21], end: [6, 20] },
    { sign: 'Cancer', start: [6, 21], end: [7, 22] },
    { sign: 'Leo', start: [7, 23], end: [8, 22] },
    { sign: 'Virgo', start: [8, 23], end: [9, 22] },
    { sign: 'Libra', start: [9, 23], end: [10, 22] },
    { sign: 'Scorpio', start: [10, 23], end: [11, 21] },
    { sign: 'Sagittarius', start: [11, 22], end: [12, 21] },
    { sign: 'Capricorn', start: [12, 22], end: [12, 31] }
  ];

  for (const { sign, start, end } of signs) {
    if (
      (month === start[0] && day >= start[1]) ||
      (month === end[0] && day <= end[1])
    ) {
      return sign;
    }
  }
  return 'Capricorn';
}

module.exports = router;

File: server/python/recommend.py
#!/usr/bin/env python3
"""
Gemstone Recommendation Engine
Scores gemstones based on zodiac sign, life goals, budget, and color preference.
Uses a weighted scoring algorithm:
  - Zodiac Match: 40%
  - Life Goals: 25%
  - Budget Fit: 20%
  - Color Preference: 15%
"""

import sys
import json
import os
from datetime import datetime


def get_zodiac_sign(dob_str):
    """Calculate zodiac sign from date of birth string (YYYY-MM-DD)."""
    try:
        dob = datetime.strptime(dob_str, "%Y-%m-%d")
    except ValueError:
        return "Unknown"

    month = dob.month
    day = dob.day

    zodiac_dates = [
        ("Capricorn", (1, 1), (1, 19)),
        ("Aquarius", (1, 20), (2, 18)),
        ("Pisces", (2, 19), (3, 20)),
        ("Aries", (3, 21), (4, 19)),
        ("Taurus", (4, 20), (5, 20)),
        ("Gemini", (5, 21), (6, 20)),
        ("Cancer", (6, 21), (7, 22)),
        ("Leo", (7, 23), (8, 22)),
        ("Virgo", (8, 23), (9, 22)),
        ("Libra", (9, 23), (10, 22)),
        ("Scorpio", (10, 23), (11, 21)),
        ("Sagittarius", (11, 22), (12, 21)),
        ("Capricorn", (12, 22), (12, 31)),
    ]

    for sign, (sm, sd), (em, ed) in zodiac_dates:
        if (month == sm and day >= sd) or (month == em and day <= ed):
            return sign

    return "Capricorn"


def score_gemstone(gemstone, zodiac_sign, goals, budget, color_preference):
    """Score a gemstone based on user preferences using weighted criteria."""
    score = 0.0

    # 1. Zodiac Match (40 points max)
    zodiac_signs_lower = [z.lower() for z in gemstone.get("zodiacSigns", [])]
    if zodiac_sign.lower() in zodiac_signs_lower:
        score += 40.0

    # 2. Life Goals Match (25 points max)
    gem_goals = [g.lower() for g in gemstone.get("goals", [])]
    if goals:
        user_goals = [g.lower() for g in goals]
        matched = sum(1 for g in user_goals if g in gem_goals)
        if len(user_goals) > 0:
            score += (matched / len(user_goals)) * 25.0

    # 3. Budget Match (20 points max)
    gem_price = gemstone.get("priceRange", "").lower()
    budget_lower = budget.lower() if budget else ""
    if gem_price == budget_lower:
        score += 20.0
    elif (budget_lower == "mid" and gem_price == "affordable") or \
         (budget_lower == "premium" and gem_price == "mid"):
        score += 10.0
    elif budget_lower == "premium" and gem_price == "affordable":
        score += 5.0

    # 4. Color Preference (15 points max)
    if color_preference:
        gem_color = gemstone.get("color", "").lower()
        pref_lower = color_preference.lower()
        if gem_color == pref_lower:
            score += 15.0
        # Partial match for similar colors
        color_families = {
            "red": ["red", "pink", "brown"],
            "blue": ["blue", "purple"],
            "green": ["green"],
            "yellow": ["yellow", "brown"],
            "white": ["white"],
            "purple": ["purple", "blue"],
            "pink": ["pink", "red"],
            "black": ["black", "brown"],
            "brown": ["brown", "yellow", "red"],
        }
        if pref_lower in color_families and gem_color in color_families[pref_lower]:
            score += 7.0

    return round(score)


def recommend(input_data):
    """Main recommendation function."""
    # Load gemstone data
    data_path = os.path.join(os.path.dirname(__file__), "..", "data", "gemstones.json")
    with open(data_path, "r", encoding="utf-8") as f:
        gemstones = json.load(f)

    # Extract user inputs
    dob = input_data.get("dob", "")
    goals = input_data.get("goals", [])
    budget = input_data.get("budget", "mid")
    color_preference = input_data.get("colorPreference", "")

    # Calculate zodiac sign
    zodiac_sign = get_zodiac_sign(dob)

    # Score all gemstones
    scored = []
    for gem in gemstones:
        gem_score = score_gemstone(gem, zodiac_sign, goals, budget, color_preference)
        gem_copy = dict(gem)
        gem_copy["matchScore"] = gem_score
        scored.append(gem_copy)

    # Sort by score descending
    scored.sort(key=lambda x: x["matchScore"], reverse=True)

    # Return top 3 recommendations
    result = {
        "zodiacSign": zodiac_sign,
        "recommendations": scored[:3]
    }

    return result


if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "No input provided"}))
        sys.exit(1)

    try:
        input_data = json.loads(sys.argv[1])
        result = recommend(input_data)
        print(json.dumps(result))
    except Exception as e:
        print(json.dumps({"error": str(e)}))
        sys.exit(1)
